packages/odyssey-react-mui/src/DataTable/useRowReordering.tsx

```diff
@@ -0,0 +1,233 @@
+
+  const odysseyDesignTokens = useOdysseyDesignTokens();
+  const dragHandleStyles = {
+    padding: odysseyDesignTokens.Spacing1,
+    borderRadius: odysseyDesignTokens.BorderRadiusMain,
+
+    "&:focus-visible": {
+      boxShadow: `0 0 0 2px ${odysseyDesignTokens.HueNeutralWhite}, 0 0 0 4px ${odysseyDesignTokens.PalettePrimaryMain}`,
+      outline: "2px solid transparent",
+      outlineOffset: "1px",
+    },
+  };
+
+  const dragHandleText = {
+    title: "Drag row or press space/enter key to start and stop reordering",
+    "aria-label":
+      "Drag row to reorder. Or, press space or enter to start and stop reordering and esc to cancel.",
+  };
+
+  const draggableTableBodyRowClassName = ({
+    currentRowId,
+    draggingRowId,
+    hoveredRowId,
+  }: {
+    currentRowId: string;
+    draggingRowId?: string;
+    hoveredRowId?: string;
+  }) => {
+    if (draggingRowId === currentRowId && hoveredRowId !== currentRowId) {
+      return "isDragging";
+    }
+
+    if (hoveredRowId === currentRowId && draggingRowId !== currentRowId) {
+      return "isDragTarget";
+    }
+
+    if (draggingRowId === currentRowId && hoveredRowId === currentRowId) {
+      return "isDragging isDragTarget";
+    }
+
+    return undefined;
+  };
+
+  const getRowFromTableAndSetHovered = (
+    table: MRT_TableInstance<MRT_RowData>,
+    id: MRT_RowData["id"],
+  ) => {
+    if (id) {
+      const nextRow: MRT_RowData = table.getRow(id);
+
+      if (nextRow) {
+        table.setHoveredRow(nextRow);
+      }
+    }
+  };
+
+  type HandleDragHandleKeyDownArgs = {
+    table: MRT_TableInstance<MRT_RowData>;
+    row: MRT_Row<MRT_RowData>;
+    event: KeyboardEvent<HTMLButtonElement>;
+  };
+
+  const handleDragHandleKeyDown = ({
+    table,
+    row,
+    event,
+  }: HandleDragHandleKeyDownArgs) => {
+    const { hoveredRow } = table.getState();
+
+    const { key } = event;
+
+    const isSpaceKey = key === " ";
+    const isEnterKey = key === "Enter";
+    const isEscapeKey = key === "Escape";
+    const isArrowDown = key === "ArrowDown";
+    const isArrowUp = key === "ArrowUp";
+    const isSpaceOrEnter = isSpaceKey || isEnterKey;
+    const zeroIndexedPageNumber = page - 1;
+    const currentIndex = row.index + zeroIndexedPageNumber * resultsPerPage;
+
+    if (isEscapeKey) {
+      resetDraggingAndHoveredRow(table);
+      return;
+    }
+
+    if (isSpaceOrEnter) {
+      event.preventDefault();
+      event.stopPropagation();
+    }
+
+    if (draggingRow) {
+      if (typeof hoveredRow?.index === "number") {
+        const { index } = hoveredRow;
+
+        if (isSpaceOrEnter) {
+          const pageRelativeIndex =
+            index + zeroIndexedPageNumber * resultsPerPage;
+
+          if (pageRelativeIndex !== currentIndex) {
+            updateRowOrder({
+              rowId: row.id,
+              newRowIndex: pageRelativeIndex,
+            });
+
+            // Can't transition CSS hover effect. Use timeout to delay hovered row effect removal
```

Comment:
I didn't write the original code here, but I believe the issue is that the row state shouldn't be changed until the CSS transition is finished.

I'm going to leave this for now; I'd rather have Bryan do any refactoring with this code, since he wrote it initially.

Reply:
IIRC correctly there was something weird about how the classes that determine which row is being 'dragged' are applied. But, I'd have to look into it again to say for sure.

Reply:
And, to be perfectly honest I was rushing this a bit as part of the keyboard navigation work so, it's possible there is a more efficient way to handle this